Fix `Bad address` when running `cmd /c [...]`

dscho opened the pull request:
In 2b4f986e49 (Cygwin: pty: Treat *.bat and *.cmd as a non-cygwin console app., 2022-07-31), we introduced a bug fix that  specifically looks for a suffix of the command's file name.

However, that file name might be set to `NULL`, namely when `null_app_name == true`, which is the case when we detected a command-line `cmd /c [...]`.

But the commit mentioned above did not account for that possibility, instead assuming that it always has to check the file name for a `.bat` or `.cmd` suffix. As a consequence, `cmd /c [...]` invocations are completely broken in v3.3.6, resulting in a `Bad address` error.

Let's guard the code properly so that it does not try to look at the file name suffix of `(WCHAR *)NULL`.

This fixes https://github.com/msys2/msys2-runtime/issues/108

dscho commented:
I [also sent this patch to Cygwin](https://inbox.sourceware.org/cygwin-patches/8rqs6n82-0oq9-2200-944n-74s7o699385o@tzk.qr/).

jeremyd2019 commented:
Thanks @dscho! I was planning to look into this, but probably wouldn't have had time until Monday

tyan0 committed:
Cygwin: pty: Fix 'Bad address' error when running 'cmd.exe /c dir'

- If the command executed is 'cmd.exe /c [...]', runpath in spawn.cc
  will be NULL. In this case, is_console_app(runpath) check causes
  access violation. This case also the command executed is obviously
  console app., therefore, treat it as console app to fix this issue.

  Addresses: https://github.com/msys2/msys2-runtime/issues/108

This is a backport of 08281cf4cc (Cygwin: pty: Fix 'Bad address' error
when running 'cmd.exe /c dir', 2022-10-22).

The original patch that was replaced by this commit had this additional
information that is now sadly lost in Cygwin's history:

In 2b4f986e49 (Cygwin: pty: Treat *.bat and *.cmd as a non-cygwin
console app., 2022-07-31), we introduced a bug fix that specifically
looks for a suffix of the command's file name.

However, that file name might be set to `NULL`, namely when
`null_app_name == true`, which is the case when we detected a
command-line `cmd /c [...]`.

But the commit mentioned above did not account for that possibility,
instead assuming that it always has to check the file name for a `.bat`
or `.cmd` suffix. As a consequence, `cmd /c [...]` invocations are
completely broken in v3.3.6, resulting in a `Bad address` error.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

lazka commented:
It was fixed slightly differently in cygwin. Which version should we use?

https://sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=commitdiff;h=08281cf4cca9593adcc3d30184322dc60fa1cd61

mati865 commented:
They are equal in functionality.

dscho commented:
> It was fixed slightly differently in cygwin. Which version should we use?
> 
> https://sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=commitdiff;h=08281cf4cca9593adcc3d30184322dc60fa1cd61

Yes, it was a _highly_ frustrating experience. @tyan0, instead of reviewing my patch and working with me so that I can fix whatever issues they find with the patch, simply replaced it, without anything in the way of communication clearly and openly (there was just a note that it was replaced, that was all).

Sadly the replacement patch [was pushed to the `cygwin-3_3-branch` already](https://github.com/cygwin/cygwin/commits/cygwin-3_3-branch), without giving anybody any chance to improve it. And there is a ton of room for improvement:
the commit message is seriously wanting, not even using the information I provided in my proposal, let alone offering any adequate explanation as to what the "issue" is that it talks about in a sentence whose grammar makes it hard to parse:

```
Cygwin: pty: Fix 'Bad address' error when running 'cmd.exe /c dir'

- If the command executed is 'cmd.exe /c [...]', runpath in spawn.cc
  will be NULL. In this case, is_console_app(runpath) check causes
  access violation. This case also the command executed is obviously
  console app., therefore, treat it as console app to fix this issue.
```

Since that commit is already set in stone by virtue of being pushed to the `cygwin-3_3-branch`, we are stuck with it.

dscho commented:
On Sat, 22 Oct 2022, Mateusz Mikuła wrote:

> They are equal in functionality.

No, they are not. In my version, we still returned `false` in case of `command.com` and `cmd.exe` being called with `/c <command>`. In @tyan0's version, it returns `true`.

mati865 commented:
Yeah, I had noticed it too late so deleted comment to avoid confusing others.

dscho commented:
Just to add insult to injury, it [looks as if the counter-proposal that was pushed directly to `cygwin-3_3-branch` is incorrect, too](https://inbox.sourceware.org/cygwin-patches/Y1ZazH6objN99mSz@calimero.vinschen.de/).
